source/extensions/filters/network/zookeeper_proxy/zookeeper_decoder.cc

```diff
@@ -0,0 +1,360 @@
+
+  // Check message length.
+  int32_t len;
+  BufferHelper::peekInt32(data, offset, len);
+  if (len < 8) {
+    callbacks_.onDecodeError();
+    return;
+  }
+
+  int32_t xid;
+  BufferHelper::peekInt32(data, offset, xid);
+
+  // "Special" requests.
+  switch (xid) {
+  case enumToIntSigned(XidCodes::CONNECT_XID):
+    parseConnect(data, offset, len);
+    return;
+  case enumToIntSigned(XidCodes::PING_XID):
+    callbacks_.onPing();
+    return;
+  case enumToIntSigned(XidCodes::AUTH_XID):
+    parseAuthRequest(data, offset, len);
+    return;
+  default:
+    break;
+  }
+
+  // "Regular" requests.
+  int32_t opcode;
+  BufferHelper::peekInt32(data, offset, opcode);
+  switch (opcode) {
+  case enumToInt(OpCodes::GETDATA):
+    parseGetDataRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::CREATE):
+    parseCreateRequest(data, offset, len, false);
+    break;
+  case enumToInt(OpCodes::CREATE2):
+    parseCreateRequest(data, offset, len, true);
+    break;
+  case enumToInt(OpCodes::SETDATA):
+    parseSetRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::GETCHILDREN):
+    parseGetChildrenRequest(data, offset, len, false);
+    break;
+  case enumToInt(OpCodes::GETCHILDREN2):
+    parseGetChildrenRequest(data, offset, len, true);
+    break;
+  case enumToInt(OpCodes::DELETE):
+    parseDeleteRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::EXISTS):
+    parseExistsRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::GETACL):
+    parseGetAclRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::SETACL):
+    parseSetAclRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::SYNC):
+    parseSyncRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::CHECK):
+    parseCheckRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::MULTI):
+    parseMultiRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::RECONFIG):
+    parseReconfigRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::SETWATCHES):
+    parseSetWatchesRequest(data, offset, len);
+    break;
+  case enumToIntSigned(OpCodes::CLOSE):
+    callbacks_.onCloseRequest();
+    break;
+  default:
+    break;
+  }
+}
+
+#define CHECK_LENGTH(LEN, MINL)                                                                    \
+  if (LEN < MINL) {                                                                                \
+    throw new EnvoyException("Package is too small");                                              \
+  }
+
+void DecoderImpl::parseConnect(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  CHECK_LENGTH(len, 28);
+
+  // Read password - skip zxid, timeout, and session id.
+  std::string passwd;
+  offset += 20;
+  BufferHelper::peekString(data, offset, passwd);
+
+  // Read readonly flag, if it's there.
+  bool readonly{};
+  try {
+    BufferHelper::peekBool(data, offset, readonly);
+  } catch (EnvoyException& e) {
```

Comment:
I would avoid using exceptions for valid code, maybe add a `peekOptionalBool` that doesn't throw on the absence of a valid bool?